src/graph.rs

```diff
@@ -232,37 +265,92 @@ impl ModuleSlot {
   }
 }
 
+/// Convert a module slot entry into a result which contains the resolved
+/// module specifier and media type or the module graph error.
+#[cfg(feature = "rust")]
+fn to_result(
+  (specifier, module_slot): (&ModuleSpecifier, &ModuleSlot),
+) -> (
+  ModuleSpecifier,
+  Result<(ModuleSpecifier, MediaType), ModuleGraphError>,
+) {
+  match module_slot {
+    ModuleSlot::Err(err) => (specifier.clone(), Err(err.clone())),
+    ModuleSlot::Module(module) => (
+      specifier.clone(),
+      Ok((module.specifier.clone(), module.media_type)),
+    ),
+    _ => (
+      specifier.clone(),
+      Err(ModuleGraphError::LoadingErr(anyhow!(
+        "Module \"{}\" is unavailable.",
+        specifier
+      ))),
+    ),
+  }
+}
+
+/// The structure which represents a module graph, which can be serialized as
+/// well as "printed".  The roots of the graph represent the "starting" point
+/// which can be located in the module "slots" in the graph. The graph also
+/// contains any redirects where the requested module specifier was redirected
+/// to another module specifier when being loaded.
 #[derive(Debug, Serialize)]
 pub struct ModuleGraph {
-  pub root: ModuleSpecifier,
+  pub roots: Vec<ModuleSpecifier>,
   #[serde(skip_serializing)]
-  maybe_locker: Option<Rc<RefCell<dyn Locker>>>,
-  #[serde(serialize_with = "serialize_modules")]
-  pub(crate) modules: BTreeMap<ModuleSpecifier, ModuleSlot>,
-  redirects: BTreeMap<ModuleSpecifier, ModuleSpecifier>,
+  maybe_locker: Option<Rc<RefCell<Box<dyn Locker>>>>,
+  #[serde(serialize_with = "serialize_modules", rename = "modules")]
+  pub(crate) module_slots: BTreeMap<ModuleSpecifier, ModuleSlot>,
+  pub redirects: BTreeMap<ModuleSpecifier, ModuleSpecifier>,
 }
 
 impl ModuleGraph {
-  pub fn new(
-    root: ModuleSpecifier,
-    maybe_locker: Option<Rc<RefCell<dyn Locker>>>,
+  fn new(
+    roots: Vec<ModuleSpecifier>,
+    maybe_locker: Option<Rc<RefCell<Box<dyn Locker>>>>,
   ) -> Self {
     Self {
-      root,
+      roots,
       maybe_locker,
-      modules: Default::default(),
+      module_slots: Default::default(),
       redirects: Default::default(),
     }
   }
 
+  /// Returns `true` if the specifier resolves to a module within a graph,
+  /// otherwise returns `false`.
+  #[cfg(feature = "rust")]
+  pub fn contains(&self, specifier: &ModuleSpecifier) -> bool {
+    let specifier = self.resolve(specifier);
+    self
+      .module_slots
+      .get(&specifier)
+      .map_or(false, |ms| matches!(ms, ModuleSlot::Module(_)))
+  }
+
+  /// Returns any errors that are in the module graph, along with the associated
+  /// specifier.
+  #[cfg(feature = "rust")]
+  pub fn errors(&self) -> Vec<(&ModuleSpecifier, &ModuleGraphError)> {
```

Comment:
Would it be better if we just stored `ModuleSpecifier` within `ModuleGraphError`? Then consumers could do something like `error.specifier()` and we could eliminate this tuple here and the usages elsewhere. I think it would make consuming code easier to understand

Side comment: never liked `error.0` and `error.1`... C# at least has named tuples, but these Rust tuples aren't ideal for a public API.

Reply:
I agree with David

Reply:
Yes, agreed.

Reply:
Done.